README.md
Apix Cache, cache-tagging for PHP
[![Build Status](https://travis-ci.org/apix/cache.svg?branch=master)](https://travis-ci.org/apix/cache)
=================================
[![Latest Stable Version](https://poser.pugx.org/apix/cache/v/stable.svg)](https://packagist.org/packages/apix/cache)
[![Total Downloads](https://poser.pugx.org/apix/cache/downloads)](https://packagist.org/packages/apix/cache)
[![Build Status](https://scrutinizer-ci.com/g/frqnck/apix-cache/badges/build.png?b=master)](https://scrutinizer-ci.com/g/frqnck/apix-cache/build-status/master)
[![Code Quality](https://scrutinizer-ci.com/g/frqnck/apix-cache/badges/quality-score.png?b=master)](https://scrutinizer-ci.com/g/frqnck/apix-cache/?branch=master)
[![Code Coverage](https://scrutinizer-ci.com/g/frqnck/apix-cache/badges/coverage.png?b=master)](https://scrutinizer-ci.com/g/frqnck/apix-cache/?branch=master)
[![License](https://poser.pugx.org/apix/cache/license.svg)](https://packagist.org/packages/apix/cache)

Apix Cache is a generic and thin cache wrapper with a PSR-6 interface to various caching backends and emphasising cache **tagging** and **indexing**.

> Cache-tagging allows to find/update all data items with one or more given tags. Providing, for instance, a batch delete of all obsolete entries matching a speficic tag such as a version string.

* Fully **unit-tested** and compliant with PSR-1, PSR-2, PSR-4 and **PSR-6** (Cache).
* [Continuously integrated](https://travis-ci.org/apix/cache)
  * against **PHP** ~~5.3~~, **5.4**, **5.5**, **5.6**, **7.0**, **7.1**, **7.2**, **7.3** ~~and HHVM~~,
  * and against `APC`, `APCu`, `Redis`, `MongoDB`, `Sqlite`, `MySQL`, `PgSQL` and `Memcached`, ...
  * supports a range of serializers: `igBinary`, `msgpack`, `json`, `php`, ...
* Extendable, additional extensions are available:
   * **[apix/simple-cache](//github.com/apix/simple-cache)** provides a SimpleCache (PSR-16) interface.
   * More contributions will be linked here.
* Available as a [Composer](https://packagist.org/packages/apix/cache) ~~and as a [PEAR](http://pear.ouarz.net)~~ package.

⇄ *[Pull requests](//github.com/frqnck/apix-cache/blob/master/.github/CONTRIBUTING.md)* and ★ *Stars* are always welcome. For bugs and feature request, please [create an issue](//github.com/frqnck/apix-cache/issues/new).

---

Cache backends
--------------
Currently, the following cache store are supplied:

* **[`APCu`](http://php.net/apcu)** and **[APC](http://php.net/apc)** *with tagging support*,
* **[Redis](#redis-specific)** using the [`PhpRedis`](https://github.com/phpredis/phpredis) extension *with tagging support*,
* **[MongoDB](#mongodb-specific)** using either the new [`mongodb`](http://php.net/mongodb) or the legacy [`mongo`](http://php.net/mongo) extension *with tagging support*,
* **[Memcached](#memcached-specific)** using the [`Memcached`](http://php.net/book.memcached.php) extension *with indexing, tagging and namespacing support*,
* and relational databases using **[PDO](#pdo-specific)** *with tagging support*:
  * Dedicated drivers for **[SQLite](http://www.sqlite.org)**, **[PostgreSQL](http://www.postgresql.org)** and **[MySQL](http://www.mysql.com)** (also works with Amazon Aurora, MariaDB and Percona),
  * A generic **[Sql1999](https://en.wikipedia.org/wiki/SQL:1999)** driver for [4D](http://www.4d.com/), [Cubrid](http://www.cubrid.org), [SQL Server](http://www.microsoft.com/sqlserver), [Sybase](http://www.sybase.com), [Firebird](http://www.firebirdsql.org), [ODBC](https://en.wikipedia.org/wiki/Open_Database_Connectivity), [Interbase](http://www.embarcadero.com/products/interbase), [IBM DB2](www.ibm.com/software/data/db2/), [IDS](http://www-01.ibm.com/software/data/informix/), [Oracle](http://www.oracle.com/database)...
* **[Directory](#filesystem-specific)** and **[Files](#filesystem-specific)** based *with tagging support*,
* **Runtime**, in-memory array storage.

Factory usage (PSR-Cache wrapper)
-------------

```php
  use Apix\Cache;

  $backend = new \Redis();
  #$backend = new \PDO('sqlite:...');    // Any supported client object e.g. Memcached, MongoClient, ...
  #$backend = new Cache\Files($options); // or one that implements Apix\Cache\Adapter
  #$backend = 'apcu';                    // or an adapter name (string) e.g. "APC", "Runtime"
  #$backend = new MyArrayObject();       // or even a plain array() or \ArrayObject.

  $cache = Cache\Factory::getPool($backend);             // without tagging support
  #$cache = Cache\Factory::getTaggablePool($backend);    // with tagging
  
  $item = $cache->getItem('wibble_id');
  
  if ( !$cache->isHit() ) {
    $data = compute_expensive_stuff();
    $item->set($data);
    $cache->save($item);
  }

  return $item->get();
```

Basic usage (Apix native)
-----------

```php
  use Apix\Cache;

  $cache = new Cache\Apcu;

  // try to retrieve 'wibble_id' from the cache
  if ( !$data = $cache->load('wibble_id') ) {
    
    // otherwise, get some data from the origin
    // example of arbitrary mixed data
    $data = array('foo' => 'bar');

    // and save it to the cache
    $cache->save($data, 'wibble_id');
  }
```
You can also use the folowing in your use cases: 
```php
  // save $data to the cache as 'wobble_id',
  // tagging it along the way as 'baz' and 'flob',
  // and set the ttl to 300 seconds (5 minutes)
  $cache->save($data, 'wobble_id', array('baz', 'flob'), 300);

  // retrieve all the cache ids under the tag 'baz'
  $ids = $cache->loadTag('baz');

  // clear out all items with a 'baz' tag
  $cache->clean('baz');

  // remove the named item
  $cache->delete('wibble_id');

  // flush out the cache (of all -your- stored items)
  $cache->flush();
```

Advanced usage
--------------
### Options shared by all the backends
```php
  use Apix\Cache;
  
  // default options
  $options = array(
      'prefix_key'  => 'apix-cache-key:', // prefix cache keys
      'prefix_tag'  => 'apix-cache-tag:', // prefix cache tags
      'tag_enable'  => true               // wether to enable tags support
  );

  // start APCu as a local cache
  $local_cache = new Cache\Apcu($options);
```

### Redis specific
```php
  // additional (default) options
  $options['atomicity']  = false;   // false is faster, true is guaranteed
  $options['serializer'] = 'php';   // null, php, igbinary, json and msgpack

  $redis_client = new \Redis;       // instantiate phpredis*
  $distributed_cache = new Cache\Redis($redis_client, $options);
```
\* see [PhpRedis](https://github.com/nicolasff/phpredis) for instantiation usage.

### Memcached specific
```php
  // additional (default) options, specific to Memcached
  $options['prefix_key'] = 'key_';  // prefix cache keys
  $options['prefix_tag'] = 'tag_';  // prefix cache tags
  $options['prefix_idx'] = 'idx_';  // prefix cache indexes
  $options['prefix_nsp'] = 'nsp_';  // prefix cache namespaces
  $options['serializer'] = 'auto';  // auto, igbinary, msgpack, php, json and json_array.

  $memcached  = new \Memcached;     // a Memcached*** instance
  $shared_cache = new Cache\Memcached($memcached, $options);
```

The serialzer `auto` (default) is `igBinary` if available, then `msgpack` if available, then `php` otherwise.

\*\*\* see [Memcached](http://php.net/manual/en/book.memcached.php) for instantiation details.

### MongoDB specific 
```php
  // additional (default) options
  $options['object_serializer'] = 'php';  // null, json, php, igBinary
  $options['db_name'] = 'apix';           // name of the mongo db
  $options['collection_name'] = 'cache';  // name of the mongo collection

  $mongo  = new \MongoDB\Client;          // MongoDB native driver
  #$mongo  = new \MongoClient;            // or MongoDB legacy driver
  $cache = new Cache\Mongo($mongo, $options);
```

### PDO specific
```php
  // additional (default) options, specific to the PDO backends
  $options['db_table']   = 'cache';       // table to hold the cache
  $options['serializer'] = 'php';         // null, php, igbinary, json and msgpack
  $options['preflight']  = true;          // wether to preflight the DB
  $options['timestamp']  = 'Y-m-d H:i:s'; // the timestamp DB format

  // with SQLITE
  $dbh = new \PDO('sqlite:/tmp/apix_tests.sqlite3');
  $relational_cache = new Cache\Pdo\Sqlite($dbh, $options);

  // with MYSQL, MariaDB and Percona
  $dbh = new \PDO('mysql:host=xxx;port=xxx;dbname=xxx', 'user', 'pass');
  $mysql_cache = new Cache\Pdo\Mysql($dbh, $options);

  // with PGSQL
  $dbh = new \PDO('pgsql:dbname=xxx;host=xxx', 'xxx', 'xxx');
  $postgres_cache = new Cache\Pdo\Pgsql($dbh, $options);

  // with a SQL:1999 compliant DB, e.g. Oracle
  $dbh = new \PDO('oci:dbname=xxx', 'xxx', 'xxx');
  $sql1999_cache = new Cache\Pdo\Sql1999($dbh, $options);
```
The `preflight` option will create on-the-fly the required tables if these are mssing.
Once these tables exist, set `preflight` to `false` in order to avoid the extraneous checks. 

### Filesystem specific

```php
  // additional (default) options
  $options['directory'] = sys_get_temp_dir() . '/apix-cache'; // Directory where the cache is created
  $options['locking'] = true;                                 // File locking (recommended)
  
  $files_cache = new Cache\Files($options);
  // or
  $directory_cache = new Cache\Directory($options);  
```

  - **Files**: the cache metadata (expiration time and tags) are stored in the cache file directly.
  - **Directory**: the metadata are stored separately from the cached data. 

Installation
------------------------

This project adheres to [Semantic Versioning](http://semver.org/) and can be installed using composer:  

    $ composer require apix/cache:^1.3

All notable changes to this project are documented in its [CHANGELOG](CHANGELOG.md).

License
-------
This work is licensed under the New BSD license -- see the [LICENSE](LICENSE.txt) for the full details.<br>Copyright (c) 2010-2016 Franck Cassedanne


CHANGELOG.md
# APIx Cache changelog

#### Version 1.3.6 (XX-Dec-2020)
- On PHP >= 7.3, if the key passed to `loadKey()` isn't in the cache a
notice is generated because value returned from `PDO::fetch()` is false
but it is treated as an array (see #40 and PR #44 by @BlairCooper).

#### Version 1.3.5 (11-Jun-2020)
- Improved the `flush` method for the SQL backends (see #40 and PR #39 by @BlairCooper).

#### Version 1.3.4 (30-Oct-2019)
- Fixed Memcached unit tests for PHP 7.x (PR #39 by @BlairCooper).
- Fixed `Memcached.php::setSerializer` for PHP 7.3 compatibility (PR #37 by @BlairCooper & PR #38 by @mgmbh).
- Fixed Travis builds with PHP 5.* (PR #37 by @BlairCooper).
- Fixed Travis on PHP 5.4 and 5.5 cannot run on Xenial (PR #37 by @BlairCooper).
- Fixed typo in composer suggested package (PR #35 by @Great-Antique).
- Added cas_token to `getTtl()` method (PR #34 contrib by @BlairCooper).
- Added check of "expire" when entry is read from a file (PR #33 contrib by @dimasikturbo).

#### Version 1.3.3 (18-May-2018)
- Modified `Redis::flush` to delete all keys only from the current DB, instead of deleting all keys from the server (PR #31 contrib by @alexpica).
- Fixes to handle changes with phpredis >= 4, cast return value to Boolean. 
- Fix `APCu` adapter to use either `\APCUIterator` or `\APCIterator`.
- Modified Travis config to handle PHP 5.3 and Precise distro.

#### Version 1.3.2 (19-Jul-2017)
- Added a dedicated `APCu` backend (+ relevant tests) as the extension no longer ship with the backward compatibility module 'apcu-bc’ (see #29).

#### Version 1.3.1 (19-Jun-2017)
- Fix a MySQL issue where same key/value returned 0 number of updated rows and triggered "SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry" (PR #28 contrib by @dimasikturbo).

#### Version 1.3.0 (2-May-2017)
- Fix a PSR-6 expiration issue.
- Added `PsrCache\Item::__toString` method to simplify cached value output.
- Added `PsrCache\Pool::__destruct()` method to (garbage collect) magically commit all deferred cached items.
- Removed all deprecated methods from `PsrCache`.
- Various fixes, more unit-tests and some cleanup.

#### Version 1.2.9 (5-Jan-2017)
- Fix `Files::clean` and `Directory::clean` return to early if failing to find a tag (PR #17 by @melloc01 + relevant tests PR #24).
- Fix `Files::flush(true)` the implementation to flush all was missing (PR #25 contrib by @alexpica).
- Updated to allow patches from php-fig/cache (PR #26 contrib by @vaibhavpandeyvpz).
- Fix to a deprecated method `PsrCache::setExpiration` (PR #27 contrib by @damianopetrungaro).

#### Version 1.2.8 (28-Oct-2016)
- Added new `mongodb` extension for PHP 5.4 and higher (contrib by @dimasikturbo) which also supports HHVM 3.9 and higher. The legacy `mongo` extension is still provided for PHP 5.6 and lower.
- Set Travis to skip `mongodb` on HHVM (compilation issue).
- Fix array serialisation of nested keys with Mongo (contrib by @dimasikturbo).

#### Version 1.2.7 (20-July-2016)
- Fix the HHVM issues.
- Fix APC/APCu for both PHP7 and HHVM.
- Updated `.travis` (optimisations).
- Added `msgpack` to Redis, Memcached and to all the PDO backends.  
- Added 'auto' and 'json_array' to Memcached.
- Changed Memcached default serializer to `auto`.
- Updated `README.md`.
- Added some additional unit-tests.
- Fix issue #15 "Files cache not correctly handling EOL on Windows" (thanks goes to @davybatsalle). 

#### Version 1.2.6 (4-July-2016)
- Fix issue #13 "TaggablePool and Pool overrides prefix_key and prefix_tag options with hardcoded value" (thanks goes to @alexpica). 
- Fix PHP 5.3, using `array()` instead of the short array syntax `[]`.
- Marcked as depreciated `isSerialized()` and `testIsSerialized()`.
- Added `msgpack` serializer.
- Set Travis to skip `Memcached` on PHP 7.0 (not yet officially supported).
- Added additional unit-tests, aiming for 100% code coverage.

#### Version 1.2.5 (20-Jun-2016)
- Fix issue #12 by adding `Files` and `Directory` backends to the Factory class (thanks goes to @alexpica). 
- Added some additional Factory tests.

#### Version 1.2.4 (6-Jan-2016)
- Updated PSR-Cache (Draft) to PSR-6 (Accepted).
- Marked as deprecated: `PsrCache::setExpiration`, `PsrCache::isRegenerating`, `PsrCache::exists`.
- Added additional unit tests to cover PSR-6.
- Updated `composer.json`.
- Updated `README.md`.
- Updated `.gitignore`.
- Added file locking option to the filesystem backends (contrib by @MacFJA).

#### Version 1.2.3 (5-Jan-2016)
- Fix APCu versions (contrib by @mimmi20).
- Added `Files` and `Directory` backends (contrib by @MacFJA).
- Updated `README.md`.

#### Version 1.2.2 (1-Sept-2015)
- Added a `CHANGELOG.md` file.
- Updated PHPUnit to 4.8 version.
- Dropped (partially) PHP 5.3 support - Memcached seems to be broken.
- Dropped PEAR support.
- Refactored `.travis.yml` tests.
- Made Travis faster (using Docker containers and skipping allowable failures).
- Added support to PHP 7.0.

#### Version 1.2.1 (4-Oct-2014)
- Added setOption().
- Updated `composer.json`.
- Updated `README.md`.
- Added Scrutinizer checks.
- Merged Scrutinizer Auto-Fixes.
- Various minor changes.

#### Version 1.2.0 (19-Sept-2014)
- Added preflight option to PDO backends.
- Added PSR-6 Cache support as a factory class.
- Added [Coverall](https://coveralls.io/github/frqnck/apix-cache) support.
- Updated `README.md`.
- Added APCu support.
- Added PHP 5.6 and HHVM support.
- Updated the `README.md`.

#### Version 1.1.0 (30-Jan-2013)
- Added `--prefer-source` (recommended by @seldaek).
- Updated the `README.md`.
- Added some unit tests.
- Added JSON support to Redis.

#### Version 1.0.5 (23-Jan-2013)
- Added `loadKey()`, `loadTag()` and removed `load()`.

#### Version 1.0.4 (22-Jan-2013)
- Added dedicated SQL1999 class.
- Fixed PDO and SQL definitions.
- Fixed `.travis.yml`. 

#### Version 1.0.3 (20-Jan-2013)
- Added some aditional tests.
- Refactored Serialisers.

#### Version 1.0.2 (20-Jan-2013)
- Added MongoDB implementation. 
- Various fixes. 


src/Memcache.php
<?php
/**
 *
 * This file is part of the Apix Project.
 *
 * (c) Franck Cassedanne <franck at ouarz.net>
 *
 * @license     http://opensource.org/licenses/BSD-3-Clause  New BSD License
 *
 */

namespace Apix\Cache;

/**
 * Memcache cache wrapper.
 *
 * @author Franck Cassedanne <franck at ouarz.net>
 *
 */
class Memcache extends AbstractCache
{
    /**
     * Holds an injected adapter.
     * @var \Memcache
     */
    protected $adapter = null;

    /**
     * Holds the array of TTLs.
     * @var array
     */
    protected $ttls = array();

    /**
     * Constructor.
     *
     * @param \Memcache $memcache A Memcache instance.
     * @param array     $options  Array of options.
     */
    public function __construct(\Memcache $memcache, array $options = null)
    {
        // default options
        $this->options['prefix_key'] = 'key_';  // prefix cache keys
        $this->options['prefix_tag'] = 'tag_';  // prefix cache tags
        $this->options['prefix_idx'] = 'idx_';  // prefix cache indexes
        $this->options['prefix_nsp'] = 'nsp_';  // prefix cache namespaces

        $this->options['serializer'] = 'php';   // none, php, json, igBinary.

        parent::__construct($memcache, $options);

        if ($this->options['tag_enable']) {
            $this->setSerializer($this->options['serializer']);
        }
    }

    /**
     * {@inheritdoc}
     */
    public function loadKey($key)
    {
        return $this->get($this->mapKey($key));
    }

    /**
     * {@inheritdoc}
     */
    public function loadTag($tag)
    {
        return $this->getIndex($this->mapTag($tag))->load();
    }

    /**
     * {@inheritdoc}
     */
    public function save($data, $key, array $tags = null, $ttl = null)
    {
        $ttl = $this->sanitiseTtl($ttl);

        $mKey = $this->mapKey($key);

        $data = array('data' => $data, 'ttl' => $ttl);
        $this->ttls[$mKey] = $ttl;

        // add the item
        $success = $this->adapter->set($mKey, $data, $ttl);

        if ($success && $this->options['tag_enable'] && !empty($tags)) {

            // add all the tags to the index key.
            $this->getIndex($this->mapIdx($key))->add($tags);

            // append the key to each tag.
            foreach ($tags as $tag) {
                $this->getIndex($this->mapTag($tag))->add($mKey);
            }
        }

        return $success;
    }

    /**
     * {@inheritdoc}
     */
    public function clean(array $tags)
    {
        $items = array();

        foreach ($tags as $tag) {
            $keys = $this->loadTag($tag);
            if (null !== $keys) {
                foreach ($keys as $key) {
                    $items[] = $key;
                }
            }

            // add the tag to deletion
            $items[] = $this->mapTag($tag);
        }

        $success = true;

        foreach ($items as $item) {
            $success = $this->adapter->delete($item) && $success;
        }

        return $success;
    }

    /**
     * {@inheritdoc}
     */
    public function delete($key)
    {
        $_key = $this->mapKey($key);
        $items = array($_key);

        if ($this->options['tag_enable']) {
            $idx_key = $this->mapIdx($key);

            // load the tags from the index key
            $tags = $this->getIndex($idx_key)->load();

            if (is_array($tags)) {
                // mark the key as deleted in the tags.
                foreach ($tags as $tag) {
                    $this->getIndex($this->mapTag($tag))->remove($_key);
                }

                // delete that index key
                $items[] = $idx_key;
            }
        }

        $success = true;

        foreach ($items as $item) {
            $success = $this->adapter->delete($item) && $success;
        }

        return $success;
    }

    /**
     * {@inheritdoc}
     */
    public function flush($all = false)
    {
        return $this->adapter->flush();
    }

    /**
     * Retrieves the cache item for the given id.
     *
     * @param  string     $id        The cache id to retrieve.
     * @return mixed|null Returns the cached data or null.
     */
    public function get($id)
    {
        $data = $this->adapter->get($id);

        if (false !== $data) {
            $this->ttls[$id] = isset($data['ttl']) ? $data['ttl'] : 0;

            return isset($data['data']) ? $data['data'] : $data;
        }

        return null;
    }

    /**
     * Returns the ttl sanitased for this cache adapter.
     *
     * The number of seconds may not exceed 60*60*24*30 = 2,592,000 (30 days).
     * @see http://php.net/manual/en/Memcache.expiration.php
     *
     * @param  integer|null $ttl The time-to-live in seconds.
     * @return int
     */
    public function sanitiseTtl($ttl)
    {
        return $ttl > 2592000 ? time() + $ttl : $ttl;
    }

    /**
     * Returns the named indexer.
     *
     * @param  string          $name The name of the index.
     * @return Indexer\Adapter
     */
    public function getIndex($name)
    {
        return new Indexer\MemcacheIndexer($name, $this);
    }

    /**
     * Returns a prefixed and sanitased cache id.
     *
     * @param  string $key The base key to prefix.
     * @return string
     */
    public function mapIdx($key)
    {
        return $this->sanitise($this->options['prefix_idx'] . $key);
    }

    /**
     * Increments the value of the given key.
     *
     * @param  string  $key The key to increment.
     *
     * @return int|bool Returns the new item's value on success or FALSE on failure.
     */
    public function increment($key)
    {
        $counter = $this->adapter->get($key);

        if (false === $counter) {
            $counter = 1;
            $this->adapter->set($key, $counter);
        } else {
            $counter = $this->adapter->increment($key);
        }

        return $counter;
    }

    /**
     * {@inheritdoc}
     *
     * The number of seconds may not exceed 60*60*24*30 = 2,592,000 (30 days).
     */
    public function getTtl($key)
    {
        $mKey = $this->mapKey($key);

        if (!isset($this->ttls[$mKey])) {
            $data = $this->adapter->get($mKey);

            $this->ttls[$mKey] = (isset($data['ttl']) ? $data['ttl'] : 0);
        }

        return $this->ttls[$mKey];
    }

    /**
     * Gets the injected adapter.
     *
     * @return \Memcache
     */
    public function getAdapter()
    {
        return $this->adapter;
    }
}
